### PR TITLE
Enhanced Clipping and Rendered Visuals Tracking in ServerCompositionVisual

### PR DESCRIPTION
**Pull request by stepangovorko**
## What does the pull request do?
This PR enhances the clipping calculation within the `ServerCompositionVisual` by taking `Clip.Bounds` into account. The primary goal is to ensure that visuals, which are not within a dirty rect, are not re-rendered when different `ClipToBounds` and `Clip` geometry parameters are set.

## What is the current behavior?
Currently, `ServerCompositionVisual.Render` method do nothing if boundaries of the visual do not intersect with `DirtyRect`. This approach effectively minimizes the areas of the scene redrawn, especially when a single element is invalidated. However, this optimization is only effective when `ClipToBounds=true`. While having `ClipToBounds=false` but setting an explicit `Clip` is valid, it's not considered in the current implementation.

## What is the updated/expected behavior with this PR?
After this PR, the `ServerCompositionVisual` will also consider changes to the `Clip` property, allowing it to recalculate when the clip is altered. This ensures accurate rendering even when `ClipToBounds=false` but an explicit `Clip` is set. 

## How was the solution implemented (if it's not obvious)?
1. `Clip.Bounds` was considered in the clipping calculation within `ServerCompositionVisual`.
2. Added unit tests under `CompositorInvalidationClippingTests` to verify the correct rendering of visuals not in the dirty rect with different `ClipToBounds` and `Clip` geometry parameters.
3. For test purpuses a counter for rendered visuals in `ServerCompositionVisual` was introduced.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0038492-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Rendering/Composition/ICompositionTargetDebugEvents.cs [hunk 1]
```diff
@@ -2,5 +2,7 @@ namespace Avalonia.Rendering.Composition;
 
 internal interface ICompositionTargetDebugEvents
 {
+    public int RenderedVisuals { get; }
```
Comment (kronic) [inline src/Avalonia.Base/Rendering/Composition/ICompositionTargetDebugEvents.cs:5]: Remove public

File: src/Avalonia.Base/Rendering/Composition/Server/ServerCompositionVisual.cs [hunk 2]
```diff
@@ -182,11 +183,24 @@ public virtual UpdateResult Update(ServerCompositionTarget root)
...
-                        .TransformToAABB(GlobalTransformMatrix)
-                    : null;
+                Rect? transformedVisualBounds = null;
+                Rect? transformedClipBounds = null;
                 
+                if (ClipToBounds)
+                    transformedVisualBounds = new Rect(new Size(Size.X, Size.Y)).TransformToAABB(GlobalTransformMatrix);
+                
+                 if (Clip != null)
+                     transformedClipBounds = Clip.Bounds.TransformToAABB(GlobalTransformMatrix);
```
Comment (kekekeks) [inline src/Avalonia.Base/Rendering/Composition/Server/ServerCompositionVisual.cs:193]: What happens if `Clip` bounds exceed control bounds?

**Review by kekekeks — APPROVED** [review]

**Comment by kekekeks** [comment]
Please address that "remove public" comment. Otherwize LGTM

**Comment by stepangovorko** [comment]
@dotnet-policy-service agree company="JetBrains GmbH"

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0038528-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0038530-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Pull request merged**